### PR TITLE
[Bug] fixing game breaking u-turn faint switch interaction

### LINKED ISSUE #2680: [BUG] U-turn incorrectly applies some effects to the switched in Pokemon
**Describe the bug**

When a Pokemon uses U-turn, some strange (and potentially game-breaking) interactions can take place:
- If the user attacks into a target with Rough Skin or Iron Barbs and faints on U-turn's last hit, the "Faint" status effect is applied to the switched in Pokemon instead of the user, causing the switched in Pokemon to also "faint" (see the video below).
- If the user attacks into a target with Flame Body, and the user would get burned on U-turn's last hit, the switched in Pokemon is burned instead of the user.

**To Reproduce**

1. Have a Pokemon with U-turn and 3x Multi-Lens (can be done via overrides)
2. Give the enemy Pokemon Rough Skin and Iron Barbs as its ability and passive (via overrides)
3. Attack into the enemy Pokemon with U-turn

**Expected behavior**

The effects listed above should be applied to U-turn's user, not its switch target.

**Screenshots / Videos**

Here's a video of U-turn's behavior when the user faints while using it:

https://github.com/pagefaultgames/pokerogue/assets/168692175/40fb9862-2c07-4eee-93d7-c4cdae02c594

**Device**

PC (Chrome)

**Additional context**

I already tried to fix this by adding another `MoveEffectTrigger` type for U-turn's attribute, but the same problems as before were observed. I think the problem is related to `FaintPhase` and `ObtainStatusEffectPhase` using battler indexes instead of the Pokemon themselves to apply effects, but I'm not sure what a clean fix to that would look like.

### PR DESCRIPTION
**Pull request by DustinLin**
## What are the changes?

The proper interaction now occurs (using pokemon now faints, and a new one is swapped in).

The dialogue boxes are still out of order from fainting, but hopefully #2743 addresses the problem.

## Why am I doing these changes?

Closes #2680 

## What did change?

Currently the `SwitchSummonPhase` is being added in the phase queues' in the wrong order for multi hit moves. It is being added too early (before the `DamagePhase`  and `FaintPhase` of the last multi-hit), which is causing the switch-in faint issue of #2680. I've written a call to add the phase right before the `MoveEndPhase` (which ensures that it wont interfere with anything of the multi-hit move, and the swapped in 'mon is present for the end-of-turn effects such as weather, abilities, field hazards).

I tried playing around with the existing functions, namely playing around with `phaseQueueSpliceIndex` to manipulate the order of the phases being added. The problem with this is that queuing ability-effects related phases also calls `setPhaseQueueSplice()`, which messes up the idea of trying to add all the phases before `SwitchSummonPhase`.

The solution I've come up with relies on the fact that `PhaseQueue` contains a `MoveEndPhase` (since it essentially does a linear search for it). I believe this should always be true, since these force-switches/damages occur during Moves. However this should be tested/confirmed.

### Screenshots/Videos

(can look at the `main` branch behavior from the linked issue)

- pr fix:

https://github.com/pagefaultgames/pokerogue/assets/39450497/ba14269c-6395-43ca-8953-de43b007accb

- interaction with weather, abilities:

https://github.com/pagefaultgames/pokerogue/assets/39450497/39bf64a4-0cfb-4968-b14f-5b5b9787d20f

- interaction with hazards:

https://github.com/pagefaultgames/pokerogue/assets/39450497/836ad2a8-83df-4a7a-89dd-f8dc15f66270

- enemy fixes as well

https://github.com/pagefaultgames/pokerogue/assets/39450497/3acd5c45-83fe-4e4c-9174-d902aba2309e

## How to test the changes?

3x Multi-Lens, and with the abilities Rough Skin and Iron Barbs gives the fainting from move effect (i used level 5 vs 15 to ensure it faints/survives)

I've only tested all the above with U-turn, we should make sure the other moves which use the ForceSwitch attribute still behave as normal

## Checklist
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [] Have I provided a clear explanation of the changes?
-[x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/battle-scene.ts [hunk 1]
```diff
@@ -2064,6 +2064,29 @@ export default class BattleScene extends SceneBase {
...
+        break;
+      }
+    }
+    if (flag === false) {
+      // what to do here if there is none?
+      // it should be the case that during combat there is always a MoveEndPhase when a move is being used, extensive testing/reasoning needed
+      this.unshiftPhase(phase);
+    }
+    return flag;
+  }
```
Comment (innerthunder) [inline src/battle-scene.ts:2088]:
```suggestion
    const targetIndex = this.phaseQueue.findIndex(ph => ph instanceof targetPhase);
    if (targetIndex !== -1) {
        this.phaseQueue.splice(targetIndex, 0, phaseToAdd);
        return true;
    } else {
        this.unshiftPhase(phaseToAdd);
        return false;
    }
  }
```

To go along with the change to the function definition (also using `findIndex` which performs better than a for loop).

File: src/data/move.ts [hunk 2]
```diff
@@ -4512,7 +4513,8 @@ export class ForceSwitchOutAttr extends MoveEffectAttr {
 	  	user.scene.triggerPokemonFormChange(switchOutTarget, SpeciesFormChangeActiveTrigger, true);
 
 	  	if (switchOutTarget.hp) {
-          user.scene.unshiftPhase(new SwitchSummonPhase(user.scene, switchOutTarget.getFieldIndex(), user.scene.currentBattle.trainer.getNextSummonIndex((switchOutTarget as EnemyPokemon).trainerSlot), false, this.batonPass, false));
+        // for opponent switching out
+          user.scene.prependToMoveEndPhase(new SwitchSummonPhase(user.scene, switchOutTarget.getFieldIndex(), user.scene.currentBattle.trainer.getNextSummonIndex((switchOutTarget as EnemyPokemon).trainerSlot), false, this.batonPass, false));
```
Comment (innerthunder) [inline src/data/move.ts:4517]:
```suggestion
          user.scene.prependToPhase(new SwitchSummonPhase(user.scene, switchOutTarget.getFieldIndex(), user.scene.currentBattle.trainer.getNextSummonIndex((switchOutTarget as EnemyPokemon).trainerSlot), false, this.batonPass, false), MoveEndPhase);
```
To go along with the change to the function definition.

File: src/field/pokemon.ts [hunk 3]
```diff
@@ -3094,7 +3094,8 @@ export class PlayerPokemon extends Pokemon {
 
       this.scene.ui.setMode(Mode.PARTY, PartyUiMode.FAINT_SWITCH, this.getFieldIndex(), (slotIndex: integer, option: PartyOption) => {
         if (slotIndex >= this.scene.currentBattle.getBattlerCount() && slotIndex < 6) {
-          this.scene.unshiftPhase(new SwitchSummonPhase(this.scene, this.getFieldIndex(), slotIndex, false, batonPass));
+          // add a SwitchSummonPhase
+          this.scene.prependToMoveEndPhase(new SwitchSummonPhase(this.scene, this.getFieldIndex(), slotIndex, false, batonPass));
```
Comment (innerthunder) [inline src/field/pokemon.ts:3098]:
```suggestion
          this.scene.prependToPhase(new SwitchSummonPhase(this.scene, this.getFieldIndex(), slotIndex, false, batonPass), MoveEndPhase);
```
To go along with the change to the function definition

File: src/battle-scene.ts [hunk 4]
```diff
@@ -2064,6 +2064,29 @@ export default class BattleScene extends SceneBase {
     }
   }
 
+  /**
+   * Tries to add the input phase to right before the MoveEndPhase, useful for resolving moves before the end-of-turn phases start
+   * @param phase {@linkcode Phase} the phase to be added
+   * @returns boolean if a MoveEndPhase was found
+   */
+  prependToMoveEndPhase(phase: Phase): boolean {
```
[the remark this answers is not in the document]
Comment (DustinLin) [reply src/battle-scene.ts]: Thanks, great suggestion

**Review by innerthunder — APPROVED** [review]

**Comment by DustinLin** [comment]
closing because PR is now pointing to `beta` and that was merged in with #2860

**Pull request closed without merging**